### PR TITLE
Add `aggregatorMergeStrategy` property in SegmentMetadata queries

### PR DESCRIPTION
**Pull request by abhishekrb19**
#### Motivation
[SegmentMetadata queries](https://druid.apache.org/docs/latest/querying/segmentmetadataquery.html) currently supports two types of aggregator merge strategies, namely strict and lenient, when "aggregators" analysis type is enabled. Users often want something less strict than a lenient policy, where the most recent aggregator is selected for a column in an evolving data model. A strict strategy is best suited once the data model is locked in.

This PR:
- Adds a new property, `aggregatorMergeStrategy`, to `segmentMetadata` queries. Please see `docs/querying/segmentmetadataquery.md` for more information.  This also allows us to define an earliest strategy (similar to the latest strategy) or more sophisticated merge strategies as needed.
- Deprecates the existing `lenientAggregatorMerge` boolean property in favor of `aggregatorMergeStrategy`.
- Validates that both the old and new property is not set. The new property is backwards compatible with the deprecated property, so the default is `strict` when aggregators analysis type is enabled. 
- Changes the segment id of merged segments as part of segmentMetadata analysis, from `merged` to  `<datasource>_<interval>_merged_<partition_number>`  format. 
- Adjusts unit tests to assert the complete `SegmentAnalysis` object instead of just the aggregators. Also, add tests for latest aggregator merge and backwards compatibility logic.

#### Release note
 `lenientAggregatorMerge` property in segment metadata queries is deprecated in favor of a new property `aggregatorMergeStrategy`. `aggregatorMergeStrategy` also supports a latest strategy in addition to existing strict and lenient strategies from `lenientAggregatorMerge`.

<hr>

##### Key changed/added classes in this PR
 * `AggregatorMergeStrategy.java`
 * `SegmentMetadataQuery.java`
 * `SegmentMetadataQueryQueryToolChest.java`
 * `SegmentMetadataQueryTest.java`
 * `SegmentMetadataQueryQueryToolChestTest.java`

<hr>

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: sql/src/test/java/org/apache/druid/sql/calcite/schema/SegmentMetadataCacheTest.java [hunk 1]
```diff
@@ -1309,7 +1310,8 @@ public void testRunSegmentMetadataQueryWithContext() throws Exception
         queryContext,
         EnumSet.noneOf(SegmentMetadataQuery.AnalysisType.class),
         false,
-        false
+        null,
+        AggregatorMergeStrategy.STRICT
```
Comment (abhishekagarwal87) [inline sql/src/test/java/org/apache/druid/sql/calcite/schema/SegmentMetadataCacheTest.java:1314]: this should be null

**Review by jon-wei — APPROVED** [review]

File: processing/src/main/java/org/apache/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java [hunk 2]
```diff
@@ -254,10 +256,10 @@ public <T extends LogicalSegment> List<T> filterSegments(SegmentMetadataQuery qu
 
   @VisibleForTesting
   public static SegmentAnalysis mergeAnalyses(
-      @Nullable String dataSource,
+      Set<String> dataSources,
```
Comment (zachjsh) [inline processing/src/main/java/org/apache/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java:259]: why use all datasources now, instead of first as before? I believe that the SegmentAnalysis is for single table?

File: processing/src/main/java/org/apache/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java [hunk 3]
```diff
@@ -309,7 +325,7 @@ public static SegmentAnalysis mergeAnalyses(
 
     final Map<String, AggregatorFactory> aggregators = new HashMap<>();
 
-    if (lenientAggregatorMerge) {
+    if (AggregatorMergeStrategy.LENIENT == aggregatorMergeStrategy) {
       // Merge each aggregator individually, ignoring nulls
       for (SegmentAnalysis analysis : ImmutableList.of(arg1, arg2)) {
```
Comment (zachjsh) [inline processing/src/main/java/org/apache/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java:330]: cant either arg1 and arg2 technically be null here based on logic above? I think that will result in exception here if so.

File: processing/src/main/java/org/apache/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java (hunk 2 above)
Comment (zachjsh) [reply processing/src/main/java/org/apache/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java:259]: Oh I see, its equivalent to before since the first datasource is checked first, and now we have added protection that if first segment didn't belong to first dataSource in list but to a subsequent one, then merge analysis will still work.

File: processing/src/main/java/org/apache/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java (hunk 3 above)
Comment (zachjsh) [reply processing/src/main/java/org/apache/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java:330]: Issue existed before, nvm

File: processing/src/main/java/org/apache/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java (hunk 2 above)
Comment (abhishekrb19) [reply processing/src/main/java/org/apache/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java:259]: yeah, this is for `UnionDataSource` - previously, the caller was passing only the first one [here](https://github.com/apache/druid/pull/14560/files#diff-acc53e0e891879963c30d5c8b349347f6b7d96c0560992a3c8fe1f9f3d2ddd3bL142). Now we pass in all the datasources and determine the datasource in this function

File: processing/src/main/java/org/apache/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java (hunk 3 above)
Comment (abhishekrb19) [reply processing/src/main/java/org/apache/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java:330]: The existing null checks at the top of the function handle this. If at least one of `arg1` and `arg2` is null, this function returns early:

```java
 if (arg1 == null) {
      return arg2;
    }
 if (arg2 == null) {
      return arg1;
}
```

File: processing/src/main/java/org/apache/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java [hunk 4]
```diff
@@ -369,7 +399,7 @@ public static SegmentAnalysis mergeAnalyses(
     if (arg1.getId() != null && arg2.getId() != null && arg1.getId().equals(arg2.getId())) {
       mergedId = arg1.getId();
     } else {
-      mergedId = "merged";
+      mergedId = mergedSegmentId == null ? "merged" : mergedSegmentId.toString();
```
Comment (zachjsh) [inline processing/src/main/java/org/apache/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java:402]: is anything else relying on this previous harcoded id which changing here could break backward compatibility? Does the `SegmentId.tryParse` function return same result for a given datasource in each case?

File: processing/src/main/java/org/apache/druid/query/metadata/metadata/SegmentMetadataQuery.java [hunk 5]
```diff
@@ -106,11 +107,28 @@ public SegmentMetadataQuery(
...
-    this.lenientAggregatorMerge = lenientAggregatorMerge == null ? false : lenientAggregatorMerge;
+    if (!(dataSource instanceof TableDataSource || dataSource instanceof UnionDataSource)) {
+      throw InvalidInput.exception("Invalid dataSource type [%s]. "
+                                   + "SegmentMetadataQuery only supports table or union datasources.", dataSource);
+    }
+    // We validate that there's only one parameter specified by the user. While the deprecated property is still
+    // supported in the API, we only set the new member variable either using old or new property, so we've a single source
+    // of truth for consumers of this class variable. The defaults are to preserve backwards compatibility.
+    // In a future release, 28.0+, we can remove the deprecated property lenientAggregatorMerge.
+    if (lenientAggregatorMerge != null && aggregatorMergeStrategy != null) {
```
Comment (zachjsh) [inline processing/src/main/java/org/apache/druid/query/metadata/metadata/SegmentMetadataQuery.java:118]: If both are specified, but consistent, should we allow?

File: processing/src/main/java/org/apache/druid/timeline/SegmentId.java [hunk 6]
```diff
@@ -231,6 +231,16 @@ static String tryExtractMostProbableDataSource(String segmentId)
...
   }
 
+
+  /**
+   * Creates a merged SegmentId for the given data source, interval and partition number. Used when segments are
+   * merged.
+   */
+  public static SegmentId merged(String dataSource, Interval interval, int partitionNum)
+  {
+    return of(dataSource, interval, "merged", partitionNum);
```
Comment (zachjsh) [inline processing/src/main/java/org/apache/druid/timeline/SegmentId.java:241]: See previous comment

File: processing/src/main/java/org/apache/druid/query/metadata/metadata/SegmentMetadataQuery.java (hunk 5 above)
Comment (abhishekrb19) [reply processing/src/main/java/org/apache/druid/query/metadata/metadata/SegmentMetadataQuery.java:118]: We could, but makes it hard to reason about

File: processing/src/main/java/org/apache/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java (hunk 3 above)
Comment (zachjsh) [reply processing/src/main/java/org/apache/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java:330]: sorry I was confusing arg1, arg2, with id1, id2. nvm. 😆

File: processing/src/main/java/org/apache/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java (hunk 4 above)
Comment (abhishekrb19) [reply processing/src/main/java/org/apache/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java:402]: Good question. I couldn't find any documentation or other parts of the system that depends on the hardcoded string "merged" - it seems like it's just treated like any other segment id and I verified that in the Druid console as well.

File: processing/src/main/java/org/apache/druid/query/metadata/metadata/SegmentMetadataQuery.java (hunk 5 above)
Comment (zachjsh) [reply processing/src/main/java/org/apache/druid/query/metadata/metadata/SegmentMetadataQuery.java:118]: true

File: processing/src/main/java/org/apache/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java (hunk 4 above)
Comment (abhishekrb19) [reply processing/src/main/java/org/apache/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java:402]: >Does the SegmentId.tryParse function return same result for a given datasource in each case?

Yes, there's some good test coverage for this [here](https://github.com/apache/druid/blob/master/processing/src/test/java/org/apache/druid/timeline/SegmentIdTest.java#L86). It'll return null if the segment id cannot be parsed for whatever reason

**Review by zachjsh — APPROVED** [review]
LGTM

File: docs/querying/segmentmetadataquery.md [hunk 7]
```diff
@@ -62,7 +62,8 @@ There are several main parts to a segment metadata query:
 |merge|Merge all individual segment metadata results into a single result|no|
 |context|See [Context](../querying/query-context.md)|no|
 |analysisTypes|A list of Strings specifying what column properties (e.g. cardinality, size) should be calculated and returned in the result. Defaults to ["cardinality", "interval", "minmax"], but can be overridden with using the [segment metadata query config](../configuration/index.md#segmentmetadata-query-config). See section [analysisTypes](#analysistypes) for more details.|no|
-|lenientAggregatorMerge|If true, and if the "aggregators" analysisType is enabled, aggregators will be merged leniently. See below for details.|no|
+|aggregatorMergeStrategy| The strategy used to merge aggregators across segments. If true, and if the "aggregators" analysisType is enabled, it defaults to "strict". Possible values include "strict", "lenient" and "latest". See [below](#aggregatormergestrategy) for details.|no|
```
Comment (ektravel) [inline docs/querying/segmentmetadataquery.md:65]:
```suggestion
|aggregatorMergeStrategy| The strategy Druid uses to merge aggregators across segments. If true and if the `aggregators` analysis type is enabled, `aggregatorMergeStrategy` defaults to `strict`. Possible values include `strict`, `lenient`, and `latest`. See [aggregatorMergeStrategy](#aggregatormergestrategy) for details.|no|
```

File: docs/querying/segmentmetadataquery.md [hunk 8]
```diff
@@ -62,7 +62,8 @@ There are several main parts to a segment metadata query:
...
+|lenientAggregatorMerge|Deprecated: Use aggregatorMergeStrategy property instead. If true, and if the "aggregators" analysisType is enabled, aggregators will be merged leniently. See below for details.|no|
```
Comment (ektravel) [inline docs/querying/segmentmetadataquery.md:66]:
```suggestion
|lenientAggregatorMerge|Deprecated. Use `aggregatorMergeStrategy` property instead. If true, and if the `aggregators` analysis type is enabled, aggregators are merged leniently.|no|
```

File: docs/querying/segmentmetadataquery.md [hunk 9]
```diff
@@ -185,7 +186,7 @@ Currently, there is no API for retrieving this information.
 * `aggregators` in the result will contain the list of aggregators usable for querying metric columns. This may be
 null if the aggregators are unknown or unmergeable (if merging is enabled).
 
-* Merging can be strict or lenient. See *lenientAggregatorMerge* below for details.
+* Merging can be `strict`, `lenient` or `latest`. See [`aggregatorMergeStrategy`](#aggregatormergestrategy) below for details.
```
Comment (ektravel) [inline docs/querying/segmentmetadataquery.md:189]:
```suggestion
* Merging can be `strict`, `lenient`, or `latest`. See [`aggregatorMergeStrategy`](#aggregatormergestrategy) for details.
```

File: docs/querying/segmentmetadataquery.md [hunk 10]
```diff
@@ -194,15 +195,20 @@ null if the aggregators are unknown or unmergeable (if merging is enabled).
 * `rollup` in the result is true/false/null.
 * When merging is enabled, if some are rollup, others are not, result is null.
 
-## lenientAggregatorMerge
+### aggregatorMergeStrategy
 
 Conflicts between aggregator metadata across segments can occur if some segments have unknown aggregators, or if
-two segments use incompatible aggregators for the same column (e.g. longSum changed to doubleSum).
+two segments use incompatible aggregators for the same column (e.g. `longSum` changed to `doubleSum`). The following
```
Comment (ektravel) [inline docs/querying/segmentmetadataquery.md:201]:
```suggestion
two segments use incompatible aggregators for the same column, such as `longSum` changed to `doubleSum`. The following
```

File: docs/querying/segmentmetadataquery.md [hunk 11]
```diff
@@ -194,15 +195,20 @@ null if the aggregators are unknown or unmergeable (if merging is enabled).
...
+aggregator merge strategies are supported:
```
Comment (ektravel) [inline docs/querying/segmentmetadataquery.md:202]:
```suggestion
Druid supports the following aggregator merge strategies:
```

File: docs/querying/segmentmetadataquery.md [hunk 12]
```diff
@@ -194,15 +195,20 @@ null if the aggregators are unknown or unmergeable (if merging is enabled).
...
 
-Aggregators can be merged strictly (the default) or leniently. With strict merging, if there are any segments
-with unknown aggregators, or any conflicts of any kind, the merged aggregators list will be `null`. With lenient
-merging, segments with unknown aggregators will be ignored, and conflicts between aggregators will only null out
-the aggregator for that particular column.
+- `strict`:  If there are any segments with unknown aggregators, or any conflicts of any kind, the merged aggregators
```
Comment (ektravel) [inline docs/querying/segmentmetadataquery.md:204]:
```suggestion
- `strict`:  If there are any segments with unknown aggregators or any conflicts of any kind, the merged aggregators
```

File: docs/querying/segmentmetadataquery.md [hunk 13]
```diff
@@ -194,15 +195,20 @@ null if the aggregators are unknown or unmergeable (if merging is enabled).
...
+  list will be `null`.
```
Comment (ektravel) [inline docs/querying/segmentmetadataquery.md:205]:
```suggestion
  list is `null`.
```

File: docs/querying/segmentmetadataquery.md [hunk 14]
```diff
@@ -194,15 +195,20 @@ null if the aggregators are unknown or unmergeable (if merging is enabled).
...
+- `lenient`:  Segments with unknown aggregators will be ignored, and conflicts between aggregators will only null out
```
Comment (ektravel) [inline docs/querying/segmentmetadataquery.md:206]:
```suggestion
- `lenient`:  Druid ignores segments with unknown aggregators. Conflicts between aggregators send the aggregator for that particular column to null.
```
I am not sure if this changes the original meaning but consider avoiding future tense and passive voice.

File: docs/querying/segmentmetadataquery.md [hunk 15]
```diff
@@ -194,15 +195,20 @@ null if the aggregators are unknown or unmergeable (if merging is enabled).
...
+- the aggregator for that particular column.
+- `latest`: In the event of conflicts between segments, the aggregator from the most recent segment will be selected
```
Comment (ektravel) [inline docs/querying/segmentmetadataquery.md:208]:
```suggestion
- `latest`: In the event of conflicts between segments, Druid selects the aggregator from the most recent segment
```

File: docs/querying/segmentmetadataquery.md [hunk 16]
```diff
@@ -194,15 +195,20 @@ null if the aggregators are unknown or unmergeable (if merging is enabled).
...
+   for that particular column.
 
-In particular, with lenient merging, it is possible for an individual column's aggregator to be `null`. This will not
-occur with strict merging.
+
+### lenientAggregatorMerge (deprecated)
+
+This property is deprecated - consider using [`aggregatorMergeStrategy`](#aggregatormergestrategy) instead.
```
Comment (ektravel) [inline docs/querying/segmentmetadataquery.md:214]:
```suggestion
Deprecated. Use [`aggregatorMergeStrategy`](#aggregatormergestrategy) instead.
```

**Review by ektravel — COMMENTED** [review]
Left a few suggestions related to the documentation.

File: docs/querying/segmentmetadataquery.md [hunk 17]
```diff
@@ -194,15 +195,20 @@ null if the aggregators are unknown or unmergeable (if merging is enabled).
...
+two segments use incompatible aggregators for the same column, such as `longSum` changed to `doubleSum`.
+Druid supports the following aggregator merge strategies:
 
-Aggregators can be merged strictly (the default) or leniently. With strict merging, if there are any segments
-with unknown aggregators, or any conflicts of any kind, the merged aggregators list will be `null`. With lenient
-merging, segments with unknown aggregators will be ignored, and conflicts between aggregators will only null out
-the aggregator for that particular column.
+- `strict`:  If there are any segments with unknown aggregators or any conflicts of any kind, the merged aggregators
+  list is `null`.
+- `lenient`:  Druid ignores segments with unknown aggregators. Conflicts between aggregators send the aggregator for that particular column to null.
```
Comment (abhishekrb19) [inline docs/querying/segmentmetadataquery.md:206]:
```suggestion
- `lenient`:  Druid ignores segments with unknown aggregators. Conflicts between aggregators set the aggregator for that particular column to null.
```

File: processing/src/main/java/org/apache/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java [hunk 18]
```diff
@@ -268,16 +270,30 @@ public static SegmentAnalysis mergeAnalyses(
...
+            (id2.getIntervalEnd().isEqual(id1.getIntervalEnd()) && id2.getPartitionNum() > id1.getPartitionNum())) {
+          mergedSegmentId = SegmentId.merged(dataSource, id2.getInterval(), id2.getPartitionNum());
+          final SegmentAnalysis tmp = arg1;
+          arg1 = arg2;
+          arg2 = tmp;
+        } else {
+          mergedSegmentId = SegmentId.merged(dataSource, id1.getInterval(), id1.getPartitionNum());
+        }
+        break;
       }
```
Comment (abhishekagarwal87) [inline processing/src/main/java/org/apache/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java:297]: what's the motivation behind this change? why will there be segment metadata queries on multiple datasources? is that something we want to support?

Comment (abhishekrb19) [reply processing/src/main/java/org/apache/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java:297]: @abhishekagarwal87, this is for Union datasources that can have multiple table sources. Segment metadata queries already [support](https://github.com/apache/druid/pull/14560/files#diff-232a53270a761b5a581cf2e575511bc7981cff11bf2f08c7bf019eb1fb310e4fR110) table and union datasources. Previously the caller was passing in only the [first](https://github.com/apache/druid/pull/14560/files#diff-acc53e0e891879963c30d5c8b349347f6b7d96c0560992a3c8fe1f9f3d2ddd3bL142) datasource to this function to align the supplied args. So `SegmentId.tryParse()` will not succeed if the datasource name wasn't the first datasource in the the union datasource.

With this change, all the table names (in the case of union datasources) are passed in to parse the segment id, sort, and align the segment analyses - `arg1` and `arg2`. I will add a Javadoc and maybe a unit test to this effect in a follow on when I add the `earliest` strategy
